### DISCUSSION TITLE
Problem with scroll restoration (nextjs 13)

### DISCUSSION THREAD

**Discussion by mikeyakymenko** [vercel/next.js#44013] [Help]
I have a hobby project, which I rebuild with nextjs, and I use 13th version. Everything actually works as expected with new app directory and I really enjoy that. I wanna go with couple production projects within 13 version, but have problem with scroll restoration, it simply doesn't works, or works incorrect. 

Image I'm on middle of home page:
<img width="1552" alt="Screenshot 2022-12-14 at 02 05 49" src="https://user-images.githubusercontent.com/887178/207489317-bf22aadf-33ad-40fe-8e6d-0d46de15e254.png">

After click on unit (team) for example 28ths Flamingos and go to the unit page:
<img width="1552" alt="Screenshot 2022-12-14 at 02 05 54" src="https://user-images.githubusercontent.com/887178/207489728-76f67a02-3c0c-496c-87b0-eebc4e572f4e.png">

so next for some reason scroll a bit, or do not scroll to top.

Expected result: 
<img width="1552" alt="Screenshot 2022-12-14 at 02 05 57" src="https://user-images.githubusercontent.com/887178/207489817-17d6e728-60e0-48e3-9349-cc70825f9cdb.png">

I used a next/link

```
<Link href={`/unit/${unit.id}`} className='flex items-center gap-1'><UnitLogo unit={unit} size={'sm'} /></Link>
```

How to fix it?

**Comment by ciruz** [comment]
Hey which next.js version are you using?

They removed some additional div's in this version [v13.0.7-canary.2](https://github.com/vercel/next.js/releases/tag/v13.0.7-canary.2). It can fix your error.

**Comment by mikeyakymenko** [comment]
@ciruz Hey thanks man, yes will `scrollRestoration: true` in `next.config.js` and [v13.0.7-canary.2](https://github.com/vercel/next.js/releases/tag/v13.0.7-canary.2) all works as expected! Thank you man!

**Reply by ciruz** [reply]
Glad I could help

**Reply by mikeyakymenko** [reply]
Hm.. I was celebrate too early. It works only in `dev` mode. In production still here. :( I've install [v13.0.7-canary.6](https://github.com/vercel/next.js/releases/tag/v13.0.7-canary.6)

**Comment by mikeyakymenko** [comment]
Any help?  I've test it with latest canary release [v13.0.8-canary.0](https://github.com/vercel/next.js/releases/tag/v13.0.8-canary.0) still there.

**Comment by mikeyakymenko** [comment]
My config file:

```
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  swcMinify: true,
  trailingSlash: true,
  generateEtags: false,
  experimental: {
    appDir: true,
    scrollRestoration: true
  }
}

module.exports = nextConfig

```

**Reply by christianledgard** [reply]
Facing the same issue here...

**Comment by AndrewIngram** [comment]
I've highlighted this issue a few times over the last few months, still seeing it today on the latest canary. I believe this is the related issue: https://github.com/vercel/next.js/issues/42492

**Reply by mikeyakymenko** [reply]
This still doesn't work even with `13.1.1`

**Comment by semy** [comment]
Is the scrollRestoration bug already fixed?

**Comment by kamami** [comment]
Any update?

**Comment by Perjan** [comment]
Still happens with 13.2.4.
Any updates? Have you tried different versions?

**Comment by stigmat4j** [comment]
Any update?

**Reply by AndrewIngram** [reply]
13.3.1 should have this fix (https://github.com/vercel/next.js/commit/48a50718ae5e8afa43227f164d6627b46f6e788a), does it help?

**Reply by stigmat4j** [reply]
@AndrewIngram wow sorry I haven't updated the minor version. Currently working

**Comment by alexPalita** [comment]
Still happens with 13.4.5...

**Comment by nmitic** [comment]
In my case I had an element (Image) positioned as sticky (tailwind) to the route I am navigating to. When I remove sticky, scroll restoration works fine, when I add back it does not, also position fixed does not work as well.

Solution for me was to add extra empty div before the sticky or fixed positioned element.

  ```
  <div>
     <div></div> empty
     <Image position="sticky" />
     <div>Rest of the content<div>
   </div>
   ```
   
   If relevant I am on version `13.4.7`

**Reply by judas-christ** [reply]
An empty div does not work for me, but a div with any content (text, nbsp etc) or min height of at least 1px does work. This is Nextjs 13.4.19.

**Comment by ranaadeel15911** [comment]
still problem scroll restoration

**Comment by taylor-lindores-reeves** [comment]
CHRIST HOW IS THIS STILL HAPPENING IN NEXTJS?

**Comment by ShiiRochi** [comment]
Seems to be fine with `"next": "^13.5.6"`

**Comment by ezeparziale** [comment]
It happens randomly in version 14.1.0

**Reply by eriknewland** [reply]
Still seems to be the case in 14.1.2

**Reply by niqwithq** [reply]
I'll only ever update next when this issue is fixed. Let's hope the day will ever come 🥲

**Reply by eriknewland** [reply]
I stand corrected! After a lot of investigation, it turned out that global styles (specifically overflow-x) issues were affecting forwards/backwards scroll restoration. Also upgraded to 14.2.2 to ensure there wasn't an issue that was solved with the latest release. 

These issues didn't quite address my concern, but shed valuable light on the problematic nature of overflow/overflow-x and scroll restoration

https://stackoverflow.com/questions/76420252/scroll-to-top-behavior-is-not-working-in-next-js-when-opening-a-new-page-via-lin

https://stackoverflow.com/questions/58008310/next-js-link-doesnt-render-with-page-scrolled-at-the-top/66316619#66316619

**Reply by ezeparziale** [reply]
In my case, I didn't use overflow-x. I saw that one user replaced Fragment with Div, and the problem was resolved.

**Comment by niqwithq** [comment]
What's the current state of affairs? I updated to 13.5.6 but the issue remains. This is really annoying.

**Comment by Arunjenson-ss** [comment]
any update on this?

**Comment by ivanstnsk** [comment]
Hey, there
This setup works for me:

1)
```typescript
// next.config.mjs

experimental: {
    scrollRestoration: true,
}
```
2)
```typescript
const router = useRouter(); // "next/navigation"
...
router.push('/mypage', { scroll: false });
```

next.js version: 14.2.3

**Reply by o1Suleyman** [reply]
is this documented anywhere?

**Reply by linobino1** [reply]
Are you sure? I tried it and for me `scrollRestoration: true` does not make any difference, and `{ scroll: false }` produces another result:
The scroll position of the previous page will be used, so if you go from page A at position 1000 to page B, then you scroll by 99 and go back to page A using `{ scroll: false}`, you will see page A at position 99.

**Comment by iTzRusso10** [comment]
Update guys?

**Comment by polluxliu** [comment]
It happens randomly in version 15.1.3. Any solutions here cannot work.

**Comment by linobino1** [comment]
scroll restoration doesn't work for me in 15.3.2

**Reply by Rifatul-Islam-Munna** [reply]
yeah same here bro, using server component still same issue, scroll not restoring when i came back to that page

**Comment by HelaGone** [comment]
Hello! Is there a workaround for this? We're seeing impact in CLS due to this behavior.

**Comment by lucasfp13** [comment]
Still not working for me in 13.5.9 :/

**Comment by dinhanhthi** [comment]
It comes again in version `15.5.4`, it scroll down randomly when switching between pages.

**Reply by Rifatul-Islam-Munna** [reply]
for me it it scroll to the top

**Reply by kacirekj** [reply]
Same here, when I click to Back in browser, it will scroll about 20px bellow. Reapeated Forward/Backward will not move the scroll any further. But the first Back click it will scroll down. At least when I'm lucky. Sometimes it's messed completely.
